---
title: "Gmnc KO quality control"
author:
   - Matthieu Moreau^[Institute of Psychiatry and Neuroscience of Paris, INSERM U1266, 75014, Paris, France, matthieu.moreau@inserm.fr] [![](https://orcid.org/sites/default/files/images/orcid_16x16.png)](https://orcid.org/0000-0002-2592-2373)
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document: 
    code_download: yes
    df_print: tibble
    highlight: haddock
    theme: cosmo
    css: "../style.css"
    toc: yes
    toc_depth: 5
    toc_float:
      collapsed: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.align = 'center', message=FALSE, warning=FALSE, cache.lazy = FALSE)
```

# Load libraries

```{r message=FALSE, warning=FALSE}
library(Seurat)
library(scrattch.hicat)
library(FateID)
library(Matrix)
library(dplyr)
library(RColorBrewer)
library(ggplot2)
library(ggExtra)
library(cowplot)
library(reticulate)
library(wesanderson)
library(princurve)
use_python("/usr/bin/python3")

#Set ggplot theme as classic
theme_set(theme_classic())
```

# Load the raw counts matrix

```{r}
Countdata <- Read10X("../../RawData/Gmnc_KO/outs/filtered_feature_bc_matrix/")

Raw.data <- CreateSeuratObject(counts = Countdata,
                              project = "Gmnc_KO",
                              min.cells = 3,
                              min.features = 800)

Raw.data$Barcodes <- rownames(Raw.data@meta.data)

rm(Countdata)

dim(Raw.data)
```
```{r}
Raw.data$percent.mito <- PercentageFeatureSet(Raw.data, pattern = "^mt-")
Raw.data$percent.ribo <- PercentageFeatureSet(Raw.data, pattern = "(^Rpl|^Rps|^Mrp)")
```

```{r}
VlnPlot(object = Raw.data, features = c("nFeature_RNA","nCount_RNA", "percent.mito", "percent.ribo"), ncol= 2) & NoAxes()
```
# Inspect cell based on relation between nUMI and nGene detected

```{r}
# Relation between nUMI and nGene detected
Cell.QC.Stat <- Raw.data@meta.data

p1 <- ggplot(Cell.QC.Stat, aes(x=nCount_RNA, y=nFeature_RNA)) + geom_point() + geom_smooth(method="lm")
p1 <- ggMarginal(p1, type = "histogram", fill="lightgrey")

p2 <- ggplot(Cell.QC.Stat, aes(x=log10(nCount_RNA), y=log10(nFeature_RNA))) + geom_point() + geom_smooth(method="lm")
p2 <- ggMarginal(p2, type = "histogram", fill="lightgrey")

plot_grid(plotlist = list(p1,p2), ncol=2, align='h', rel_widths = c(1, 1)) ; rm(p1,p2)
```

Cells with deviating nGene/nUMI ratio display an Erythrocyte signature 


```{r}
Raw.data <- AddModuleScore(Raw.data,
                           features = list(c("Hbb-bt", "Hbq1a", "Isg20", "Fech", "Snca", "Rec114")),
                           ctrl = 10,
                           name = "Erythrocyte.signature")

Cell.QC.Stat$Erythrocyte.signature <- Raw.data$Erythrocyte.signature1
```

```{r}
gradient <- colorRampPalette(brewer.pal(n =11, name = "Spectral"))(100)

p1 <- ggplot(Cell.QC.Stat, aes(log10(nCount_RNA), y=log10(nFeature_RNA))) +
      geom_point(aes(color= Erythrocyte.signature))  + 
      scale_color_gradientn(colours=rev(gradient), name='Erythrocyte score') + theme(legend.position="none")

p2 <- ggplot(Cell.QC.Stat, aes(log10(nCount_RNA), y=log10(nFeature_RNA))) +
      geom_point(aes(color= percent.mito))  + 
      scale_color_gradientn(colours=rev(gradient), name='Percent mito') + theme(legend.position="none")

p3 <- ggplot(Cell.QC.Stat, aes(log10(nCount_RNA), y=log10(nFeature_RNA))) +
      geom_point(aes(color= percent.ribo))  + 
      scale_color_gradientn(colours=rev(gradient), name='Percent ribo') + theme(legend.position="none")

p1 + p2 + p3
```
## Exclude Erythrocytes

```{r}
Cell.QC.Stat$Erythrocyte <- ifelse(Cell.QC.Stat$Erythrocyte.signature > 0.1, "Erythrocyte", "Not_Erythrocyte")
```

```{r}
p2 <- ggplot(Cell.QC.Stat, aes(x=log10(nCount_RNA), y=log10(nFeature_RNA))) +
  geom_point(aes(colour = Erythrocyte)) +
  theme(legend.position="none")

ggMarginal(p2, type = "histogram", fill="lightgrey")
```

```{r}
# Filter cells based on these thresholds
Cell.QC.Stat <- Cell.QC.Stat %>% filter(Cell.QC.Stat$Erythrocyte.signature < 0.1)
```

# Low quality cell filtering

## Filtering cells based on percentage of mitochondrial transcripts

We applied a high and low median absolute deviation (mad) thresholds to exclude outlier cells

```{r}
max.mito.thr <- median(Cell.QC.Stat$percent.mito) + 3*mad(Cell.QC.Stat$percent.mito)
min.mito.thr <- median(Cell.QC.Stat$percent.mito) - 3*mad(Cell.QC.Stat$percent.mito)
```

```{r}
p1 <- ggplot(Cell.QC.Stat, aes(x=nFeature_RNA, y=percent.mito)) +
  geom_point() +
  geom_hline(aes(yintercept = max.mito.thr), colour = "red", linetype = 2) +
  geom_hline(aes(yintercept = min.mito.thr), colour = "red", linetype = 2) +
  annotate(geom = "text", label = paste0(as.numeric(table(Cell.QC.Stat$percent.mito > max.mito.thr | Cell.QC.Stat$percent.mito < min.mito.thr)[2])," cells removed\n",
                                         as.numeric(table(Cell.QC.Stat$percent.mito > max.mito.thr | Cell.QC.Stat$percent.mito < min.mito.thr)[1])," cells remain"),
           x = 6000, y = 20)

ggMarginal(p1, type = "histogram", fill="lightgrey", bins=100) 
```
```{r}
# Filter cells based on these thresholds
Cell.QC.Stat <- Cell.QC.Stat %>% filter(percent.mito < max.mito.thr) %>% filter(percent.mito > min.mito.thr)
```

## Filtering cells based on number of genes and transcripts detected

### Remove cells with to few gene detected or with to many UMI counts

We filter cells which are likely to be doublet based on their higher content of transcript detected as well as cell with to few genes/UMI sequenced

```{r}
# Set low and hight thresholds on the number of detected genes based on the one obtain with the WT dataset
min.Genes.thr <- log10(1635)
max.Genes.thr <- log10(8069)

# Set hight threshold on the number of transcripts
max.nUMI.thr <- log10(58958)
```


```{r}
# Gene/UMI scatter plot before filtering
p1 <- ggplot(Cell.QC.Stat, aes(x=log10(nCount_RNA), y=log10(nFeature_RNA))) +
  geom_point() +
  geom_smooth(method="lm") +
  geom_hline(aes(yintercept = min.Genes.thr), colour = "green", linetype = 2) +
  geom_hline(aes(yintercept = max.Genes.thr), colour = "green", linetype = 2) +
  geom_vline(aes(xintercept = max.nUMI.thr), colour = "red", linetype = 2)

ggMarginal(p1, type = "histogram", fill="lightgrey")
```
```{r}
# Filter cells base on both metrics
Cell.QC.Stat <- Cell.QC.Stat %>% filter(log10(nFeature_RNA) > min.Genes.thr) %>% filter(log10(nCount_RNA) < max.nUMI.thr)
```

### Filter cells below the main population nUMI/nGene relationship

```{r}
lm.model <- lm(data = Cell.QC.Stat, formula = log10(nFeature_RNA) ~ log10(nCount_RNA))

p2 <- ggplot(Cell.QC.Stat, aes(x=log10(nCount_RNA), y=log10(nFeature_RNA))) +
  geom_point() +
  geom_smooth(method="lm") +
  geom_hline(aes(yintercept = min.Genes.thr), colour = "green", linetype = 2) +
  geom_hline(aes(yintercept = max.Genes.thr), colour = "green", linetype = 2) +
  geom_vline(aes(xintercept = max.nUMI.thr), colour = "red", linetype = 2) +
  annotate(geom = "text", label = paste0(dim(Cell.QC.Stat)[1], " QC passed cells"), x = 4, y = 3.8)

ggMarginal(p2, type = "histogram", fill="lightgrey")
```

## Filter the Seurat object

```{r}
Raw.data <- subset(x = Raw.data, subset = Barcodes %in%  Cell.QC.Stat$Barcodes)
```

```{r}
# Plot final QC metrics
VlnPlot(object = Raw.data, features = c("nFeature_RNA","nCount_RNA", "percent.mito", "percent.ribo"), ncol= 2) & NoAxes()
```
```{r}
p1 <- ggplot(Raw.data@meta.data, aes(x=log10(nCount_RNA), y=log10(nFeature_RNA))) + geom_point() + geom_smooth(method="lm")
ggMarginal(p1, type = "histogram", fill="lightgrey")
```
```{r}
rm(list = ls()[!ls() %in% "Raw.data"])
```


# Use Scrublet to detect obvious doublets

## Run Scrublet with default parameter

Export raw count matrix as input to Scrublet

```{r}
#Export filtered matrix
dir.create("../../RawData/Gmnc_KO/Scrublet_inputs")

exprData <- Matrix(as.matrix(Raw.data@assays[["RNA"]]@counts), sparse = TRUE)
writeMM(exprData, "../../RawData/Gmnc_KO/Scrublet_inputs/matrix1.mtx")
```
```{python}
import scrublet as scr
import scipy.io
import numpy as np
import os

#Load raw counts matrix and gene list
input_dir = '../../RawData/Gmnc_KO/Scrublet_inputs'
counts_matrix = scipy.io.mmread(input_dir + '/matrix1.mtx').T.tocsc()

#Initialize Scrublet
scrub = scr.Scrublet(counts_matrix,
                     expected_doublet_rate=0.1,
                     sim_doublet_ratio=2,
                     n_neighbors = 8)

#Run the default pipeline
doublet_scores, predicted_doublets = scrub.scrub_doublets(min_counts=1, 
                                                          min_cells=3, 
                                                          min_gene_variability_pctl=85, 
                                                          n_prin_comps=25)
```

```{r}
# Import scrublet's doublet score
Raw.data$Doubletscore <- py$doublet_scores

# Plot doublet score
ggplot(Raw.data@meta.data, aes(x = Doubletscore, stat(ndensity))) +
  geom_histogram(bins = 200, colour ="lightgrey")+
  geom_vline(xintercept = 0.15, colour = "red", linetype = 2)
```
```{r}
# Manually set threshold at doublet score to 0.2
Raw.data$Predicted_doublets <- ifelse(py$doublet_scores > 0.15, "Doublet","Singlet")
table(Raw.data$Predicted_doublets)
```
```{r}
Raw.data <- subset(x = Raw.data, subset = Predicted_doublets == "Singlet")
```

# Generate SRING dimentionality reduction

## Export counts matrix

```{r}
dir.create("./SpringCoordinates")
```

```{r}
# Export raw expression matrix and gene list to regenerate a spring plot
exprData <- Matrix(as.matrix(Raw.data@assays[["RNA"]]@counts), sparse = TRUE)
writeMM(exprData, "./SpringCoordinates/ExprData.mtx")
```

```{r}
Genelist <- row.names(Raw.data@assays[["RNA"]]@counts)
write.table(Genelist, "./SpringCoordinates/Genelist.csv", sep="\t", col.names = F, row.names = F, quote = F)
```

```{r}
#Export metadata
Scrublet <- c("Scrublet", Raw.data$Predicted_doublets)
Scrublet <- paste(Scrublet, sep=",", collapse=",")

Cellgrouping <- Scrublet
write.table(Cellgrouping, "./SpringCoordinates/Cellgrouping.csv", quote =F, row.names = F, col.names = F)
```

## Import coordinates

```{r}
spring.coor <- read.table("SpringCoordinates/coordinates.txt", sep = ",", header = F, row.names = 1)
colnames(spring.coor) <- c("Spring_1", "Spring_2")
```

```{r}
Spring.Sym <- function(x){
  x = abs(max(spring.coor$Spring_2)-x)
 }

spring.coor$Spring_2 <- sapply(spring.coor$Spring_2, function(x) Spring.Sym(x))
```

```{r}
Raw.data$Spring_1 <- spring.coor$Spring_1
Raw.data$Spring_2 <- spring.coor$Spring_2
```


```{r}
spring <- as.matrix(Raw.data@meta.data %>% select("Spring_1", "Spring_2"))
  
Raw.data[["spring"]] <- CreateDimReducObject(embeddings = spring, key = "Spring_", assay = DefaultAssay(Raw.data))
```

```{r}
DimPlot(Raw.data, 
        reduction = "spring",
        pt.size = 0.5) & NoAxes()
```
# Broad clustering

## Sctransform normalization

```{r class.output="scroll-100", cache=TRUE}
Raw.data <- SCTransform(Raw.data,
                        method = "glmGamPoi",
                        vars.to.regress = c("percent.mito"),
                        verbose = T)
```
## Run PCA and broad clustering

```{r class.output="scroll-100", cache=TRUE}
Raw.data <- RunPCA(Raw.data, verbose = FALSE)

Raw.data <- FindNeighbors(Raw.data,
                          dims = 1:20,
                          k.param = 8)

Raw.data <- FindClusters(Raw.data, resolution = 0.2)
```

```{r}
DimPlot(Raw.data,
        reduction = "spring",
        cols = c("#ebcb2e", "#9ec22f", "#a9961b", "#cc3a1b", "#d14c8d", "#4cabdc", "#5ab793", "#e7823a", "#046c9a", "#4990c9"),
        pt.size = 0.5) & NoAxes()
```
```{r}
Raw.data$Broadclust.ident <- Raw.data$seurat_clusters
```

# Differentiating neurons sub-clustering

## Extract differentiating neurons

```{r}
Neurons.data <-  subset(Raw.data, idents = 3)

DimPlot(Neurons.data,
        reduction = "spring",
        pt.size = 1,
        cols =  c("#cc3a1b")) + NoAxes()
```
## Fit pseudotime

```{r}
fit <- principal_curve(as.matrix(Neurons.data@meta.data[,c("Spring_1", "Spring_2")]),
                       smoother='lowess',
                       trace=TRUE,
                       f = 1,
                       stretch=0)
```

```{r}
#Pseudotime score
PseudotimeScore <- fit$lambda/max(fit$lambda)

if (cor(PseudotimeScore, Neurons.data@assays$SCT@data['Hmga2', ]) > 0) {
  Neurons.data$PseudotimeScore <- -(PseudotimeScore - max(PseudotimeScore))
}

cols <- brewer.pal(n =11, name = "Spectral")

ggplot(Neurons.data@meta.data, aes(Spring_1, Spring_2)) +
  geom_point(aes(color=PseudotimeScore), size=2, shape=16) + 
  scale_color_gradientn(colours=rev(cols), name='Pseudotime score')
```
# Late Neurons diversity

## Extract late neurons

```{r}
Neurons.data$Cell.state <- cut(Neurons.data$PseudotimeScore,
                              c(0,0.4,0.8,1),
                              include.lowest = T,
                              labels=c("BP","EN","LN"))
```

```{r}
DimPlot(Neurons.data,
        group.by = "Cell.state",
        reduction = "spring",
        cols = c("#ebcb2e", "#9ec22f", "#a9961b"),
        pt.size = 1.5) & NoAxes()
```
```{r}
LN.data <- subset(Neurons.data, subset = Cell.state == "LN")
```

```{r}
DimPlot(LN.data,
        reduction = "spring",
        cols = c("#ebcb2e", "#9ec22f", "#a9961b"),
        pt.size = 1.5) & NoAxes()
```
## Prepare the dataset for clustering with scrattch.hicat

### Gene filtering

```{r}
# Exclude genes detected in less than 3 cells
num.cells <- Matrix::rowSums(LN.data@assays[["RNA"]]@counts > 0)
genes.use <- names(x = num.cells[which(x = num.cells >= 3)])

GenesToRemove <- c(grep(pattern = "(^Rpl|^Rps|^Mrp)", x = genes.use, value = TRUE),
                   grep(pattern = "^mt-", x = genes.use, value = TRUE),
                   "Xist")

genes.use <- genes.use[!genes.use %in% GenesToRemove]
```

### Normalization

```{r}
dgeMatrix_count <- as.matrix(LN.data@assays[["RNA"]]@counts)[rownames(LN.data@assays[["RNA"]]@counts) %in% genes.use,]
dgeMatrix_cpm <- cpm(dgeMatrix_count)
norm.dat <- log2(dgeMatrix_cpm + 1)

norm.dat <- Matrix(norm.dat, sparse = TRUE)
Data.matrix <- list(raw.dat=dgeMatrix_count, norm.dat=norm.dat)
attach(Data.matrix)
```

### Exclude unwanted sources of variation

```{r}
gene.counts <- log2(colSums(as.matrix(Data.matrix$norm.dat) > 0))
nUMI <- log2(colSums(Data.matrix$raw.dat))
perctMito <- LN.data$percent.mito
perctRibo <- LN.data$percent.ribo
Pseudotime <- LN.data$PseudotimeScore

rm.eigen <- as.matrix(cbind(gene.counts,
                            nUMI,
                            perctMito,
                            perctRibo,
                            Pseudotime))

row.names(rm.eigen) <- names(gene.counts)

colnames(rm.eigen) <- c("log2nGenes",
                        "log2nUMI",
                        "perctMito",
                        "perctRibo",
                        "Pseudotime ")

rm(gene.counts, nUMI, perctMito, perctRibo, Pseudotime)
```

## Iterative clustering

```{r}
# Parameters for iterative clustering
de.param <- de_param(padj.th     = 0.01, 
                     lfc.th      = 0.9,
                     low.th      = 1, 
                     q1.th       = 0.25, 
                     q2.th       = NULL,
                     q.diff.th   = 0.7,
                     de.score.th = 80,
                     min.cells   = 10)
```


```{r class.output="scroll-100", cache=TRUE}
iter.result <- iter_clust(norm.dat, 
                          counts = raw.dat,
                          dim.method = "pca",
                          max.dim = 15,
                          k.nn = 8,
                          de.param = de.param,
                          rm.eigen = rm.eigen,
                          rm.th = 0.7,
                          vg.padj.th = 0.5,
                          method = "louvain",
                          prefix = "test-iter_clust",
                          verbose = F)
```

```{r}
# Merge clusters which are not seperable by DEGs
rd.dat <- t(norm.dat[iter.result$markers,])
merge.result <- merge_cl(norm.dat, 
                         cl = iter.result$cl, 
                         rd.dat = rd.dat,
                         de.param = de.param)

cat(length(unique(merge.result$cl))," Clusters")
```
```{r}
LN.data$iter.clust <- merge.result$cl

Idents(LN.data) <- "iter.clust"

colors <-  c("#ebcb2e", "#9ec22f", "#cc3a1b")

DimPlot(LN.data,
        reduction = "spring",
        #cols = colors,
        pt.size = 1.5) & NoAxes()
```

```{r class.output="scroll-100"}
Neurons.markers <- FindAllMarkers(LN.data,
                                  test.use = "roc",
                                  only.pos = TRUE,
                                  min.pct = 0.25,
                                  logfc.threshold = 0.25)
```
```{r}
top10 <- Neurons.markers %>%
          group_by(cluster) %>%
          filter(power > 0.45)

DoHeatmap(LN.data,
          group.colors = c("#ebcb2e", "#9ec22f", "#cc3a1b"),
          features = top10$gene) + NoLegend()
```
```{r}
FeaturePlot(object = LN.data,
            features = c("Foxg1", "Zfpm2",
                         "Lhx1", "Zic5", "Zfp503"),
            pt.size = 1,
            cols = c("grey90", brewer.pal(9,"YlGnBu")),
            reduction = "spring",
            order = T) & NoAxes() & NoLegend()
```
## Use fate ID to infer lineages along differentiating cells

```{r}
Neurons.data$Broadclust.ident <- sapply(Neurons.data$Barcodes,
                              FUN = function(x) {
                                if (x %in% LN.data$Barcodes) {
                                  x = paste0("Neuron_", LN.data@meta.data[x, "iter.clust"])
                                } else {
                                  x = Neurons.data@meta.data[x, "Broadclust.ident"]
                                  }
                              })

Idents(Neurons.data) <- "Broadclust.ident"
```

```{r}
DimPlot(Neurons.data,
        reduction = "spring",
        cols = c("#ebcb2e", "#9ec22f", "#a9961b", "#cc3a1b", "#d14c8d", "#4cabdc", "#5ab793", "grey90", "#e7823a", "#046c9a", "#4990c9", "grey60"),
        pt.size = 0.5) & NoAxes()
```

## Run FateID

### FateID

```{r class.output="scroll-100", cache=TRUE}
Neurons.data <- SCTransform(Neurons.data,
                            method = "glmGamPoi",
                            vars.to.regress = c("percent.mito", "percent.ribo"),
                            verbose = T)

Neurons.data <- FindVariableFeatures(Neurons.data, selection.method = "vst", nfeatures = 1500)
```

```{r}
Norm.Mat <- as.data.frame(as.matrix(Neurons.data@assays$SCT@data[Neurons.data@assays$SCT@var.features,]))

#Rename idents
id <- 4:1
names(id) <- levels(Neurons.data)
Neurons.data <- RenameIdents(Neurons.data, id)

# Set a cluster assignment factor for each cells
ClusterIdent <- Idents(Neurons.data)
names(ClusterIdent) <- names(Idents(Neurons.data))

Attractors <- 1:3

# Distance in spring space
z <- as.matrix(dist(cbind(Neurons.data$Spring_1, Neurons.data$Spring_2)))
```

```{r class.output="scroll-100", cache=TRUE}
Infered.Fate.bias  <- fateBias(Norm.Mat, ClusterIdent, Attractors,
                               z = z,
                               minnr=20,
                               minnrh=30,
                               adapt=TRUE,
                               confidence=0.75,
                               nbfactor=5,
                               use.dist=FALSE,
                               seed=1234,
                               nbtree=NULL)
```

### Inspect test set used iteratively

```{r}
Neurons.data$FateID.iteration <- "Attractors"
Idents(Neurons.data) <- "FateID.iteration"

for (i in seq(0, length(Infered.Fate.bias$rfl), by = 5)[-1]) {
  iter <- seq(i-4,i)
  Barcodes <- c()
  for (j in iter) {
    Barcodes <- c(Barcodes, names(Infered.Fate.bias$rfl[[j]]$test$predicted))
  }
  Neurons.data <- SetIdent(Neurons.data, cells = Barcodes, value = paste0("iter ",iter[1]," to ", iter[4]))
}

DimPlot(Neurons.data,
        reduction = "spring",
        pt.size = 1) & NoAxes()
```

### Import lineage bias into Seurat meta.data

```{r}
probs <- Infered.Fate.bias$probs[,seq(length(Attractors))]

Neurons.data$prob.1 <- probs$t1
Neurons.data$prob.2 <- probs$t2
Neurons.data$prob.3 <- probs$t3

FeaturePlot(object = Neurons.data,
            features = c("prob.1", "prob.2", "prob.3"),
            pt.size = 0.5,
            cols = rev(RColorBrewer::brewer.pal(n = 11, name = "Spectral")),
            reduction = "spring",
            order = T) & NoAxes() & NoLegend()
```
```{r}
New.data <- data.frame(barcode=Neurons.data$Barcodes,
                       cluster= Neurons.data$Broadclust.ident,
                       spring1= Neurons.data$Spring_1,
                       spring2= Neurons.data$Spring_2,
                       prob.1= Neurons.data$prob.1,
                       prob.2= Neurons.data$prob.2,
                       prob.3 = Neurons.data$prob.3)

New.data$lineage.bias <- colnames(New.data[,5:7])[apply(New.data[,5:7],1,which.max)]

ggplot(New.data, aes(spring1, spring2, colour = lineage.bias)) +
  scale_color_manual(values=c("#e7823a","#cc391b","#026c9a","#d14c8d")) +
  geom_point() 
```

# Transfert ident to the full dataset

```{r}
Neurons.data$Lineage.bias <- New.data$lineage.bias

Raw.data$Broadclust.ident <- sapply(Raw.data$Barcodes,
                              FUN = function(x) {
                                if (x %in% Neurons.data$Barcodes) {
                                  x = paste0("Neuron_",Neurons.data@meta.data[x, "Lineage.bias"])
                                } else {
                                  x = Raw.data@meta.data[x, "Broadclust.ident"]
                                  }
                              })

Idents(Raw.data) <- "Broadclust.ident"
```


```{r}
DimPlot(Raw.data,
        reduction = "spring",
        cols = c("#ebcb2e", "#9ec22f", "#a9961b", "#cc3a1b", "#d14c8d", "#4cabdc", "#5ab793", "#e7823a", "#046c9a", "grey90", "#4990c9"),
        pt.size = 0.5) & NoAxes()
```
```{r}
rm(list = ls()[!ls() %in% "Raw.data"])
gc()
```
# Project progenitors domain ident from WT

```{r}
WT.KO <- list(WT = readRDS("../QC.filtered.clustered.cells.RDS") %>%
                subset(subset = orig.ident == "Hem1" & Cell_ident %in% c("ChP_progenitors", "ChP",
                                                                         "Dorso-Medial_pallium", "Medial_pallium",
                                                                         "Hem", "Thalamic_eminence") ),
              KO = Raw.data %>% subset(idents = c(1,2,3,5)))

```


```{r}
p1 <- DimPlot(object = WT.KO[["WT"]],
        group.by = "Cell.state",
        reduction = "spring",
        cols = c("#31b6bd", "#ebcb2e", "#9ec22f", "#cc3a1b", "#d14c8d", "#4cabdc", "#5ab793", "#e7823a", "#046c9a", "#4990c9"),
        pt.size = 1.5
        )  & NoAxes()

p2 <- DimPlot(WT.KO[["KO"]],
        reduction = "spring",
        group.by = "Broadclust.ident",
        cols = c("#ebcb2e", "#9ec22f", "#a9961b", "#cc3a1b"),
        pt.size = 1.5) & NoAxes()

p1 + p2
```
```{r}
WT.KO[["WT"]] <- NormalizeData(WT.KO[["WT"]], normalization.method = "LogNormalize", scale.factor = 10000, assay = "RNA")
WT.KO[["KO"]] <- NormalizeData(WT.KO[["KO"]], normalization.method = "LogNormalize", scale.factor = 10000, assay = "RNA")
```

```{r}
WT.KO[["WT"]] <- FindVariableFeatures(WT.KO[["WT"]], selection.method = "vst", nfeatures = 2000)
WT.KO[["KO"]] <- FindVariableFeatures(WT.KO[["KO"]], selection.method = "vst", nfeatures = 2000)
```

```{r}
features <- SelectIntegrationFeatures(object.list = WT.KO, nfeatures = 1500)

TFs <- read.table("TF.csv", sep = ";")[,1]
TFs <- features[features %in% TFs]
```

## transfert identity labels WT to KO

```{r class.output="scroll-100", cache=TRUE}
KO.anchors <- FindTransferAnchors(reference = WT.KO[["WT"]],
                                  query = WT.KO[["KO"]],
                                  features = TFs,
                                  reduction = "rpca",
                                  k.anchor = 5,
                                  k.filter = 100,
                                  k.score = 30,
                                  npcs = 25,
                                  dims = 1:25,
                                  max.features = 200)

predictions <- TransferData(anchorset = KO.anchors,
                            refdata = WT.KO[["WT"]]$Cell.state,
                            dims = 1:25)

WT.KO[["KO"]] <- AddMetaData(WT.KO[["KO"]], metadata = predictions)
```
```{r}
cols <- brewer.pal(n =11, name = "Spectral")

ggplot(WT.KO[["KO"]]@meta.data, aes(Spring_1, Spring_2)) +
  geom_point(aes(color=prediction.score.max), size=1, shape=16) + 
  scale_color_gradientn(colours=rev(cols), name='prediction.score.max')
```

```{r}
p1 <- DimPlot(object = WT.KO[["WT"]],
        group.by = "Cell.state",
        reduction = "spring",
        cols = c("#7293c8", "#b79f0b", "#3ca73f","#31b6bd",
                 "#ebcb2e", "#9ec22f", "#a9961b", "#cc3a1b",
                 "#d14c8d", "#4cabdc", "#5ab793", "#e7823a",
                 "#046c9a", "#4990c9"),
        pt.size = 1)  & NoAxes()

p2 <- DimPlot(WT.KO[["KO"]],
              group.by = "predicted.id",
              reduction = "spring",
              cols = c("#31b6bd", "#ebcb2e", "#9ec22f", "#a9961b", "#cc3a1b"),
              pt.size = 1) & NoAxes()

p1 + p2
```

## Transfert to the full dataset

```{r}
Raw.data$Cell.ident <- sapply(Raw.data$Barcodes,
                              FUN = function(x) {
                                if (x %in% WT.KO[["KO"]]$Barcodes) {
                                  x = WT.KO[["KO"]]@meta.data[x, "predicted.id"]
                                } else {
                                  x = Raw.data@meta.data[x, "Broadclust.ident"]
                                  }
                              })
```


```{r}
DimPlot(object = Raw.data,
        group.by = "Cell.ident",
        reduction = "spring",
        cols = c( "#4cabdc", "#7293c8", "grey40" ,"#3ca73f","grey80",
                  "#31b6bd", "#ebcb2e", "#9ec22f", "#a9961b",
                 "#046c9a", "#cc3a1b","#4990c9","#e7823a"),
        pt.size = 0.5)  & NoAxes()
```

# Change gene name annotation

Reads were realigned using the same transcriptome annotation as the WT E11.5-E12.5 dataset. We re import the count matrix from the realignment.

```{r}
rm(list = ls()[!ls() %in% "Raw.data"])
gc()
```


```{r class.output="scroll-100", cache=TRUE}
Raw.data@assays[["RNA"]]@counts <- Read10X("../../RawData/Gmnc_KO/mm10_ref/outs/filtered_feature_bc_matrix/")[,Raw.data$Barcodes]
Raw.data@assays[["RNA"]]@data <- Raw.data@assays[["RNA"]]@counts

Raw.data <- SCTransform(Raw.data,
                        method = "glmGamPoi",
                        vars.to.regress = c("percent.mito"),
                        verbose = T)
```

# Save the object

```{r Save RDS}
saveRDS(Raw.data, "./GmncKO.cells.RDS")
```


# Session Info

```{r}
#date
format(Sys.time(), "%d %B, %Y, %H,%M")

#Packages used
sessionInfo()
```