---
title: "Analysis of pallial and hem progenitors cycling dynamics using revelio"
author:
   - Matthieu Moreau^[Institute of Psychiatry and Neuroscience of Paris, INSERM U1266, 75014, Paris, France, matthieu.moreau@inserm.fr] [![](https://orcid.org/sites/default/files/images/orcid_16x16.png)](https://orcid.org/0000-0002-2592-2373)
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document: 
    code_download: yes
    df_print: tibble
    highlight: haddock
    theme: cosmo
    css: "../style.css"
    toc: yes
    toc_depth: 5
    toc_float:
      collapsed: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.align = 'center', message=FALSE, warning=FALSE)

# To use biomart 
new_config <- httr::config(ssl_verifypeer = FALSE)
httr::set_config(new_config, override = FALSE)
```

In this part of the analysis we apply [Revelio](https://github.com/danielschw188/Revelio) algorithm to explore cell cycle dynamic of pallial and hem domain radial glial cells

# Load libraries

```{r message=FALSE, warning=FALSE}
library(Seurat)
library(Revelio)
library(orthologsBioMART)
library(Matrix)
library(dplyr)
library(RColorBrewer)
library(ggplot2)
library(ggExtra)
library(cowplot)
library(wesanderson)

#Set ggplot theme as classic
theme_set(theme_classic())
```

# Load and filter progenitors data

```{r}
Hem.data <- readRDS("../QC.filtered.cells.RDS")
```

```{r}
DimPlot(object = Hem.data,
        group.by = "Cell_ident",
        reduction = "spring",
        cols = c("#83c3b8", #"ChP"
                 "#009fda", #"ChP_progenitors"
                 "#68b041", #"Dorso-Medial_pallium"
                 "#e46b6b", #"Hem"
                 "#e3c148", #"Medial_pallium"
                 "#b7d174", #2
                 "grey40", #4
                 "black", #5
                 "#3e69ac" #"Thalamic_eminence"
                 ))
```
We first test the tool on the hem progenitor subset

```{r}
Idents(Hem.data) <- Hem.data$Cell_ident
```


```{r}
Progenitors.data <-  subset(Hem.data, idents = c("Dorso-Medial_pallium", "Hem","Medial_pallium"))

DimPlot(Progenitors.data,
        reduction = "spring",
        pt.size = 1,
        cols =  c("#68b041", "#e3c148", "#e46b6b")) + NoAxes()
```

# Prepare data for revelio input

## Find mous ortologues to provided human cell cycle genes

```{r}
Cellcyclegenes <- revelioTestData_cyclicGenes
head(Cellcyclegenes)
```

We use [orthologsBioMART](https://vitkl.github.io/orthologsBioMART/index.html) library to map human to mouse mouse orthologs

```{r cache=TRUE}
G1.S <- findOrthologsHsMm(from_filters = "hgnc_symbol",
                          from_values = as.character(Cellcyclegenes$G1.S), 
                          to_attributes = "external_gene_name")

S <- findOrthologsHsMm(from_filters = "hgnc_symbol",
                          from_values = as.character(Cellcyclegenes$S), 
                          to_attributes = "external_gene_name")

G2 <- findOrthologsHsMm(from_filters = "hgnc_symbol",
                          from_values = as.character(Cellcyclegenes$G2), 
                          to_attributes = "external_gene_name")

G2.M <- findOrthologsHsMm(from_filters = "hgnc_symbol",
                          from_values = as.character(Cellcyclegenes$G2.M), 
                          to_attributes = "external_gene_name")

M.G1 <- findOrthologsHsMm(from_filters = "hgnc_symbol",
                          from_values = as.character(Cellcyclegenes$M.G1), 
                          to_attributes = "external_gene_name")


gene.list <- list(G1.S$external_gene_name,
                  S$external_gene_name,
                  G2$external_gene_name,
                  G2.M$external_gene_name,
                  M.G1$external_gene_name)

CCgenes <- t(plyr::ldply(gene.list, rbind))

colnames(CCgenes) <- colnames(Cellcyclegenes)
```


## Export counts matrix

```{r}
rawCounts <- as.matrix(Progenitors.data[["RNA"]]@counts)
```

```{r}
# Filter genes expressed by less than 10 cells
num.cells <- Matrix::rowSums(rawCounts > 0)
genes.use <- names(x = num.cells[which(x = num.cells >= 10)])
rawCounts <- rawCounts[genes.use, ]
```

```{r}
rm(list = ls()[!ls() %in% c("rawCounts", "CCgenes", "Progenitors.data")])
gc()
```

# Run Revelio


We can now follow the tutorial form the [package github page](https://github.com/danielschw188/Revelio) 

```{r cache=TRUE}
myData <- createRevelioObject(rawData = rawCounts,
                              cyclicGenes = CCgenes,
                              lowernGeneCutoff = 0,
                              uppernUMICutoff = Inf,
                              ccPhaseAssignBasedOnIndividualBatches = F)

rm("rawCounts")
gc()
```

The `getCellCyclePhaseAssignInformation` filter "outlier" cells for cell cycle phase assignation. We modified the function to keep all cells as we observed this does not affect the global cell cycle fitting procedure


```{r cache=TRUE}
source("../Functions/functions_InitializationCCPhaseAssignFiltering.R")

myData <- getCellCyclePhaseAssign_allcells(myData)
```
```{r include=FALSE}
rm(list = ls()[!ls() %in% c("Progenitors.data", "myData")])
gc()
```


```{r cache=TRUE}
myData <- getPCAData(dataList = myData)
myData <- getOptimalRotation(dataList = myData)
```
# Compare results between different ordering approched applied to the DC space

## Phase assigment projection

```{r}
Revelio.phase <- as.factor(sapply(Progenitors.data$Barcodes,
                                           function(x) {
                                             if (x %in% myData@cellInfo$cellID) {
                                               x = as.character(myData@cellInfo[x,"ccPhase"])
                                             } else {x = "Outlier"}
                                             }))

levels(Revelio.phase) <- c("G1.S", "S", "G2", "G2.M", "M.G1","Outlier")

Progenitors.data$Revelio.phase <- Revelio.phase

table(Progenitors.data$Revelio.phase)
```


```{r}
DimPlot(object = Progenitors.data,
        group.by = "Revelio.phase",
        reduction = "spring",
        cols =  c(wes_palette("FantasticFox1"),"grey90"),
        split.by = 'Revelio.phase')
```

## Comparison of Revelio cell cycle ordering and *Angle* method

To obtain the best linear ordering of cell along G1 to M phase we compare the output `ccPercentageUniformlySpaced' form Revelio and the one obtained with the [*Angle*](https://github.com/dynverse/ti_angle) method describe in [Saelens et. al. Nature Biotechnology 2019](https://www.nature.com/articles/s41587-019-0071-9)

New dataframe 

```{r}
CellCycledata <- cbind(as.data.frame(t(myData@transformedData$dc$data[1:2,])),
                       nUMI= myData@cellInfo$nUMI,
                       Revelio.phase = factor(myData@cellInfo$ccPhase, levels = c("G1.S", "S", "G2", "G2.M", "M.G1")),
                       Revelio.cc= myData@cellInfo$ccPercentageUniformlySpaced,
                       Seurat.cc= Progenitors.data@meta.data[myData@cellInfo$cellID,"CC.Difference"])
```

Compute cycling trajectory with Angle 

```{r}
CellCycledata$Angle.cc <- atan2(CellCycledata$DC1, CellCycledata$DC2) / 2 / pi + .5
```

## Cells distribution in the DC1-DC2 space

```{r}
p1 <- ggplot(CellCycledata, aes(DC1, DC2)) +
        geom_point(aes(color = Revelio.phase)) +
        scale_color_manual(values= c(wes_palette("FantasticFox1")[1:3],"grey40",wes_palette("FantasticFox1")[5]))

p2 <- ggplot(CellCycledata, aes(DC1, DC2)) +
        geom_point(aes(color=Seurat.cc), size=2, shape=16) + 
        scale_color_gradientn(colours=rev(colorRampPalette(brewer.pal(n =11, name = "Spectral"))(100)),
                              name='Seurat_cc')

p3 <- ggplot(CellCycledata, aes(DC1, DC2)) +
        geom_point(aes(color=Angle.cc), size=2, shape=16) + 
        scale_color_gradientn(colours=rev(colorRampPalette(brewer.pal(n =11, name = "Spectral"))(100)),
                              name='Angle.cc')

p4 <- ggplot(CellCycledata, aes(DC1, DC2)) +
        geom_point(aes(color=Revelio.cc), size=2, shape=16) + 
        scale_color_gradientn(colours=rev(colorRampPalette(brewer.pal(n =11, name = "Spectral"))(100)),
                              name='Revelio_cc')


(p1+p2)/(p3+p4)
```
Cycling trajectory obtain with Revelio seems to start with G2.M / M.G1 phase while the one computed with Angle starts from G1.S

## Cells distribution along Revelio and Angle computed trajectories

```{r message=FALSE, warning=FALSE}
p1 <- ggplot(CellCycledata, aes(x=Angle.cc, y=Revelio.cc, color= Revelio.phase)) +
        geom_point() +
        scale_color_manual(values= c(wes_palette("FantasticFox1")[1:3],"grey40",wes_palette("FantasticFox1")[5]))

p2 <- ggplot(CellCycledata, aes(x=Revelio.cc, y = Revelio.phase, fill= Revelio.phase)) +
        ggridges::geom_density_ridges() +
        scale_fill_manual(values= c(wes_palette("FantasticFox1")[1:3],"grey40",wes_palette("FantasticFox1")[5]))

p3 <- ggplot(CellCycledata, aes(x=Angle.cc, y = Revelio.phase, fill= Revelio.phase)) +
        ggridges::geom_density_ridges() +
        scale_fill_manual(values= c(wes_palette("FantasticFox1")[1:3],"grey40",wes_palette("FantasticFox1")[5]))

p1/p2/p3
```
## Cellular transcript content along the two computed axis

However, when looking at cells' UMI content along the cycling trajectory, Revelio's ordering appears to properly order cells from M.G1 to G2.M since the 1/2 drop in UMI content append at the end of the trajectory

- Mean UMI per .05 interval along the *Angle* axis

```{r}
meanUMI <- CellCycledata %>%
            mutate(Angle.interval = cut(CellCycledata$Angle.cc, seq(0, max(CellCycledata$Angle.cc) + 0.05, 0.05), dig.lab = 2, right = FALSE)) %>%
            select(Angle.interval, nUMI) %>%
            group_by(Angle.interval) %>%
            summarise(meanUMI =mean(nUMI))

meanUMI$max.Revelio.phase <- CellCycledata %>%
            mutate(Angle.interval = cut(CellCycledata$Angle.cc, seq(0, max(CellCycledata$Angle.cc) + 0.05, 0.05), dig.lab = 2, right = FALSE)) %>%
            select(Angle.interval, Revelio.phase) %>%
            group_by(Angle.interval, Revelio.phase) %>%
            summarise(nb = n()) %>%
            filter(nb == max(nb)) %>%
            pull(Revelio.phase)

p1 <- ggplot(meanUMI, aes(x=Angle.interval, y=meanUMI/10000, fill=max.Revelio.phase)) +
        geom_bar(stat = "identity", width = 0.90) +
        theme(axis.text.x = element_text(angle = 45, hjust=1))+
        scale_fill_manual(values= c(wes_palette("FantasticFox1")[1:3],"grey40",wes_palette("FantasticFox1")[5]))
```

- Mean UMI per .05 interval along the *Revelio* axis

```{r}
meanUMI <- CellCycledata %>%
            mutate(Revelio.interval = cut(CellCycledata$Revelio.cc, seq(0, max(CellCycledata$Revelio.cc) + 0.05, 0.05), dig.lab = 2, right = FALSE)) %>%
            select(Revelio.interval, nUMI) %>%
            group_by(Revelio.interval) %>%
            summarise(meanUMI =mean(nUMI))

meanUMI$max.Revelio.phase <- CellCycledata %>%
            mutate(Revelio.interval = cut(CellCycledata$Revelio.cc, seq(0, max(CellCycledata$Revelio.cc) + 0.05, 0.05), dig.lab = 2, right = FALSE)) %>%
            select(Revelio.interval, Revelio.phase) %>%
            group_by(Revelio.interval, Revelio.phase) %>%
            summarise(nb = n()) %>%
            filter(nb == max(nb)) %>%
            pull(Revelio.phase)

p2 <- ggplot(meanUMI, aes(x=Revelio.interval, y=meanUMI/10000, fill=max.Revelio.phase)) +
        geom_bar(stat = "identity", width = 0.90) +
        theme(axis.text.x = element_text(angle = 45, hjust=1))+
        scale_fill_manual(values= c(wes_palette("FantasticFox1")[1:3],"grey40",wes_palette("FantasticFox1")[5]))
```


```{r}
p1 / p2
```
Base on this last plot where we interpret the UMI content drop as corresponding to the cytokinesis at the end of the cell cycle, we decided to use the Revelio computed axis for further cell cycle analysis

# Save cycling trajectories into Seurat metadata slot

```{r}
Progenitors.data$Revelio.DC1 <- CellCycledata$DC1
Progenitors.data$Revelio.DC2 <- CellCycledata$DC2

Progenitors.data$Angle.cc <- CellCycledata$Angle.cc
Progenitors.data$Revelio.cc <- CellCycledata$Revelio.cc
```

# Save Progenitors seurat obj

```{r}
saveRDS(Progenitors.data,"Progenitors.RDS")
```


# Session Info

```{r}
#date
format(Sys.time(), "%d %B, %Y, %H,%M")

#Packages used
sessionInfo()
```